lib/utils/useWeb3Modal/index.tsx

```diff
@@ -1,141 +1,129 @@
 import { useEffect, useState } from "react";
-import { ethers } from "ethers";
+import { providers } from "ethers";
 import { urls } from "../../constants";
 import {
   web3InitialState,
   ConnectedWeb3State,
   Web3ModalState,
   Web3State,
   TypedProvider,
+  WrappedProvider,
+  TorusProvider,
+  CoinbaseProvider,
+  WalletConnectProvider,
 } from "../../components/Web3Context/types";
 
+/** Type guards for convenience and readability */
+const isTorusProvider = (p?: WrappedProvider): p is TorusProvider =>
+  !!p && "isTorus" in p && p.isTorus;
+
+const isWalletConnectProvider = (
+  p?: WrappedProvider
+): p is WalletConnectProvider =>
+  !!p && "isWalletConnect" in p && p.isWalletConnect;
+
+const isCoinbaseProvider = (p?: WrappedProvider): p is CoinbaseProvider =>
+  !!p && "isCoinbaseWallet" in p && p.isCoinbaseWallet;
+
+const getWeb3Provider = (p: any): TypedProvider => {
+  return new providers.Web3Provider(p) as TypedProvider;
+};
+
 /** React Hook to create and manage the web3Modal lifecycle */
 export const useWeb3Modal = (): Web3ModalState => {
   const [web3state, setWeb3State] = useState<Web3State>(web3InitialState);
 
   const disconnect = async () => {
     localStorage.removeItem("web3-wallet");
-    if (web3state && (web3state.provider?.provider as any)?.isTorus === true) {
-      await (web3state.provider?.provider as any).torus.cleanUp(); // this will trigger reload via accountsChanged
+    if (isTorusProvider(web3state.provider?.provider)) {
+      await web3state.provider?.provider?.torus?.cleanUp();
     } else if (
-      web3state &&
-      web3state.provider?.connection?.url === "metamask"
+      isWalletConnectProvider(web3state.provider?.provider) ||
+      isCoinbaseProvider(web3state.provider?.provider)
     ) {
-      window.location.reload();
-    } else {
-      (web3state.provider?.provider as any).close();
-      window.location.reload();
+      await web3state.provider?.provider?.close();
     }
+    window.location.reload();
```

Comment:
@goatsOnDrugs refactored this
torus -> `cleanUp()`
coinbase and walletconnect -> `close()`
everything else -> just reload